File: src/Chapter04GraphBFS/SingleSourcePath.java
package Chapter04GraphBFS;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;

public class SingleSourcePath {

    private Graph G;
    private int s;

    private boolean[] visited;
    private int[] pre;

    public SingleSourcePath(Graph g, int s) {
        G = g;
        s = s;
        visited = new boolean[G.V()];
        pre = new int[G.V()];
        for (int i = 0; i < pre.length; i++)
            pre[i] = -1;
        bfs(s);
    }

    public void bfs(int s) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(s);
        visited[s] = true;
        pre[s] = s;
        while (!queue.isEmpty()) {
            int v = queue.remove();

            for (int w : G.adj(v))
                if (!visited[w]) {
                    queue.add(w);
                    visited[w] = true;
                    pre[w] = v;
                }
        }
    }

    public boolean isConnectedTo(int t) {
        G.validateVertex(t);
        return visited[t];
    }

    public Iterable<Integer> path(int t) {

        ArrayList<Integer> res = new ArrayList<Integer>();
        if (!isConnectedTo(t)) return res;

        int cur = t;
        while (cur != s) {
            res.add(cur);
            cur = pre[cur];
        }
        res.add(s);

        Collections.reverse(res);
        return res;
    }

    public static void main(String[] args) {

        Graph g = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g.txt");
        SingleSourcePath sspath = new SingleSourcePath(g, 0);
        System.out.println("0 -> 6 : " + sspath.path(6));
    }

}

File: src/Chapter04GraphBFS/python/SingleSourcePath.py
from Chapter02GraphDFS.python.adjSet import adjSet as Graph
from collections import deque


class SingleSourcePath:
    def __init__(self, G, s):
        self._G = G
        self._visited = [False] * G.V
        self._pre = [-1] * G.V
        self.s = s
        for v in range(G.V):
            if not self._visited[v]:
                self._bfs(v)

    def _bfs(self, s):
        queue = deque()
        queue.append(s)
        self._pre[s] = s
        self._visited[s] = True

        while queue:
            v = queue.popleft()
            for w in self._G.adj(v):
                if not self._visited[w]:
                    queue.append(w)
                    self._visited[w] = True
                    self._pre[w] = v

    def path(self, w):
        self._G._validate_vertex(w)
        res = [w]
        while self._pre[w] != w:
            res.append(self._pre[w])
            w = self._pre[w]

        return list(reversed(res))

    def is_connected_to(self, t):
        self._G._validate_vertex(t)
        # if t is visited for current self._s
        # that implies t is in the path of self._s
        return self._visited[t]


if __name__ == "__main__":
    filename = "/Chapter03GraphDFSApplications/g.txt"
    g = Graph(filename)

    sspath = SingleSourcePath(g, 0)
    print('0 -> 6 : {}'.format(sspath.path(6)))
